### PR TITLE
Compare QuantumScript  classes with qml.equal

### PR DESCRIPTION
**Pull request by minhtriet**
**Context:**
Implement comparison between QuantumScript classes with qml.equal

**Description of the Change:**

**Benefits:**
One unified interface to compare things together, now support QuantumScript

**Possible Drawbacks:**

**Related GitHub Issues:** https://github.com/PennyLaneAI/pennylane/issues/4894

### CODE REVIEW

**Comment by minhtriet** [comment]
`tests/tape/test_qscript.py` are identical and `pennylane/ops/functions/equal.py` are mainly different in the last lines. I think PyCharm changes the spaces inside these files

**Comment by minhtriet** [comment]
@albi3ro :)

**Comment by minhtriet** [comment]
ready for round 2 @albi3ro :)

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -400,4 +402,4 @@ Mudit Pandey,
 Matthew Silverman,
 Jay Soni,
 David Wierichs,
-Justin Woodring,
+Justin Woodring, Minh Chau
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:405]: Feel free to move yourself to just below Astral Cai (it's alphabetic based on last name) :smile:

File: pennylane/ops/functions/equal.py [hunk 2]
```diff
@@ -168,6 +169,38 @@ def _equal(
...
+    op2: qml.tape.QuantumScript,
+    check_interface=True,
+    check_trainability=True,
+    rtol=1e-5,
+    atol=1e-9,
+):
+    # operations
+    if len(op1.operations) != len(op2.operations):
+        return False
+    for comparands in itertools.zip_longest(op1.operations, op2.operations):
```
Comment (mudit2812) [inline pennylane/ops/functions/equal.py:184]: Why has `itertools.zip_longest` been used here? If you've already validated that the length operations is the same for both scripts, then you can just use the vanilla python `zip`

```suggestion
    for comparands in zip(op1.operations, op2.operations):
```

Comment (minhtriet) [reply pennylane/ops/functions/equal.py]: fixed in 6a4d2097aed42790c12ca6420a7c20381c45b8d6

File: pennylane/ops/functions/equal.py [hunk 3]
```diff
@@ -168,6 +169,38 @@ def _equal(
...
+            comparands[0],
+            comparands[1],
+            check_interface=check_interface,
+            check_trainability=check_trainability,
+            rtol=rtol,
+            atol=atol,
+        ):
+            return False
+    # measurements
+    if op1.measurements != op2.measurements:
```
Comment (mudit2812) [inline pennylane/ops/functions/equal.py:194]: I would suggest using a loop and also adding checks for interfaces and trainability for the measurements. Observables such as `Hamiltonian`, `Hermitian`, `Sum`, `Prod` can have numerical parameters.

File: doc/releases/changelog-dev.md (hunk 1 above)
Comment (minhtriet) [reply doc/releases/changelog-dev.md]: fixed in 3bb4b7cb7bf5e7ef5125fe0c831ec4c552646132

Comment (minhtriet) [reply doc/releases/changelog-dev.md]: feel free to resolve this @trbromley :)

File: pennylane/ops/functions/equal.py (hunk 2 above)
Comment (minhtriet) [reply pennylane/ops/functions/equal.py]: feel free to resolve this @mudit2812 :)

File: pennylane/ops/functions/equal.py (hunk 3 above)
Comment (minhtriet) [reply pennylane/ops/functions/equal.py]: feel free to resolve this @mudit2812 :)

**Review by mudit2812 — APPROVED** [review]
Looks good to me! Thanks @minhtriet for the contribution 🚀

**Review by albi3ro — APPROVED** [review]
I left an optional comment on the docstring, but other than that I'm happy to approve. 
I'm really glad to get this behaviour 🎉

**Comment by minhtriet** [comment]
The `codecov/project` test seemed to freeze, but https://app.codecov.io/gh/PennyLaneAI/pennylane/pull/4902 said CI Passed

**Pull request merged**